.devcontainer/dev/dev.Dockerfile

```diff
@@ -0,0 +1,109 @@
+FROM ghcr.io/ubcsailbot/sailbot_workspace/base:ros_humble-ompl_d8375d8-amd64 as dev
+
+# From https://github.com/athackst/dockerfiles/blob/32a872348af0ad25ec4a6e6184cb803357acb6ab/ros2/humble.Dockerfile
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && apt-get install -y \
+  bash-completion \
+  build-essential \
+  cmake \
+  gdb \
+  git \
+  pylint \
+  python3-argcomplete \
+  python3-colcon-common-extensions \
+  python3-pip \
+  python3-rosdep \
+  python3-vcstool \
+  vim \
+  wget \
+  # Install ros distro testing packages
+  ros-humble-ament-lint \
+  ros-humble-launch-testing \
+  ros-humble-launch-testing-ament-cmake \
+  ros-humble-launch-testing-ros \
+  python3-autopep8 \
+  && rm -rf /var/lib/apt/lists/* \
+  && rosdep init || echo "rosdep already initialized" \
+  # Update pydocstyle
+  && pip install --upgrade pydocstyle
+
+ARG USERNAME=ros
+ARG USER_UID=1000
+ARG USER_GID=$USER_UID
+
+# Create a non-root user
+RUN groupadd --gid $USER_GID $USERNAME \
+  && useradd -s /bin/bash --uid $USER_UID --gid $USER_GID -m $USERNAME \
+  # [Optional] Add sudo support for the non-root user
+  && apt-get update \
+  && apt-get install -y sudo \
+  && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME\
+  && chmod 0440 /etc/sudoers.d/$USERNAME \
+  # Cleanup
+  && rm -rf /var/lib/apt/lists/* \
+  && echo "source /usr/share/bash-completion/completions/git" >> /home/$USERNAME/.bashrc \
+  && echo "if [ -f /opt/ros/${ROS_DISTRO}/setup.bash ]; then source /opt/ros/${ROS_DISTRO}/setup.bash; fi" >> /home/$USERNAME/.bashrc
+
+ENV DEBIAN_FRONTEND=
+ENV AMENT_CPPCHECK_ALLOW_SLOW_VERSIONS=1
+
+FROM dev
+
+ARG USERNAME=ros
+ARG HOME=/home/$USERNAME
+
+# persist ROS logs
+RUN mkdir -p ${HOME}/.ros/log \
+    && chown -R ${USERNAME} ${HOME}/.ros
+
+# persist bash history
+RUN SNIPPET="export PROMPT_COMMAND='history -a' && export HISTFILE=${HOME}/commandhistory/.bash_history" \
+  && mkdir ${HOME}/commandhistory \
+  && touch ${HOME}/commandhistory/.bash_history \
+  && chown ${USERNAME} ${HOME}/commandhistory/.bash_history \
+  && echo $SNIPPET >> "${HOME}/.bashrc"
+
+ARG ROS_WORKSPACE=/workspaces/sailbot_workspace
+
+# bash configuration
+COPY update-bashrc.sh /sbin/update-bashrc
+RUN chmod +x /sbin/update-bashrc ; chown ros /sbin/update-bashrc ; sync ; /bin/bash -c /sbin/update-bashrc ; rm /sbin/update-bashrc
+
+# install clang, some clang tools, and protobuf for network systems
+# TODO: move required packages to base image
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && \
+  apt-get install -y \
+    clang \
+    clang-tidy \
+    clangd \
+    googletest \
+    libboost-all-dev \
+    libprotobuf-dev \
+    llvm \
+    protobuf-compiler && \
+  apt-get autoremove -y && \
+  apt-get clean -y && \
+  rm -rf /var/lib/apt/lists/{apt,dpkg,cache,log} /tmp/* /var/tmp/*
+ENV DEBIAN_FRONTEND=
```

Comment:
@hhenry01 How does this look to you? Will the apt version of boost and googletest suffice, they're much newer in 22.04 than 18.04. See the changelog section of this PR's description for more details.

Also, which packages are necessary for functionality? We should migrate them to the base image.

Reply:
Getting a cmake error, is the best course of action to downgrade cmake or upgrade boost?
![image](https://user-images.githubusercontent.com/65756895/203696299-f1af58e9-84e8-417f-be2f-fd4711fd3008.png)

Reply:
Also is it worth installing the cmake tools extension?
![image](https://user-images.githubusercontent.com/65756895/203696702-86286aa3-4c9c-4372-8e96-1eca44ba7bea.png)

Reply:
> Also is it worth installing the cmake tools extension? ![image](https://user-images.githubusercontent.com/65756895/203696702-86286aa3-4c9c-4372-8e96-1eca44ba7bea.png)

Nope. It serves two functions:
1. Make starting a new CMake based project easier - already done
2. Make managing different builds and running them easier - we already have scripts/tasks to do this.

Reply:
> Getting a cmake error when building, is the best course of action to downgrade cmake or upgrade boost? ![image](https://user-images.githubusercontent.com/65756895/203696424-2b683da5-257e-4791-8db3-2cea8ac3e3ca.png)

Darn, I was hoping apt would have v1.78 or higher. If it simplifies things then we can use Boost v1.74

Reply:
When possible I think we should use apt repositories. If you want the features offered by newer version we can use them. To save space in the image we can build it in one container then copy over the binaries to the main container. See `.devcontainer/base/base.Dockerfile` for some examples

Reply:
Also is it necessary to install `libbost-all-dev` in the dev image? Here is what is installed for ompl in the base image
![image](https://user-images.githubusercontent.com/65756895/204072556-0b60666f-96fc-4576-8bc1-c710a2ef0fd0.png)

Reply:
I'm ok with the version of boost on apt. Does it take much longer to install `libboost-all-dev`? Finding out later down the line that a specific boost library is not installed and having to install it in the image would be quite inconvenient.

Reply:
> I'm ok with the version of boost on apt. 

- I created a PR for this: https://github.com/UBCSailbot/network_systems/pull/8

> Does it take much longer to install `libboost-all-dev`? Finding out later down the line that a specific boost library is not installed and having to install it in the image would be quite inconvenient.

- Should be fine

Also, which packages are necessary for functionality? We should migrate them from dev to base image.

```
    clang \
    clang-tidy \
    clangd \
    googletest \
    libboost-all-dev \
    libprotobuf-dev \
    llvm \
    protobuf-compiler && \
```

Reply:
Still not able to build network_systems in this branch. I updated `LLVM_VERSION` to 14 in network_systems CMakeLists.txt, but now I get this:

```
Starting >>> network_systems
--- stderr: network_systems                             
Building Network Systems with build type 'RelWithDebInfo' and flags: '-O2 -g -DNDEBUG -std=c++14 -Wall -Wextra -Wpedantic -Werror -Wno-unused-parameter -pthread'
gmake[2]: *** No rule to make target '/usr/local/lib/libgtest.a', needed by 'projects/example/test_example'.  Stop.
gmake[2]: *** Waiting for unfinished jobs....
error: too many errors emitted, stopping now [clang-diagnostic-error]
/opt/ros/humble/include/rclcpp/rclcpp/client.hpp:550:10: error: no member named 'optional' in namespace 'std' [clang-diagnostic-error]
    std::optional<CallbackInfoVariant>
    ~~~~~^
/opt/ros/humble/include/rclcpp/rclcpp/client.hpp:550:19: error: use of undeclared identifier 'CallbackInfoVariant' [clang-diagnostic-error]
    std::optional<CallbackInfoVariant>
                  ^
...
```

Reply:
> Still not able to build network_systems in this branch. I updated `LLVM_VERSION` to 14 in network_systems CMakeLists.txt, but now I get this:
> 
> ```
> Starting >>> network_systems
> --- stderr: network_systems                             
> Building Network Systems with build type 'RelWithDebInfo' and flags: '-O2 -g -DNDEBUG -std=c++14 -Wall -Wextra -Wpedantic -Werror -Wno-unused-parameter -pthread'
> gmake[2]: *** No rule to make target '/usr/local/lib/libgtest.a', needed by 'projects/example/test_example'.  Stop.
> gmake[2]: *** Waiting for unfinished jobs....
> error: too many errors emitted, stopping now [clang-diagnostic-error]
> /opt/ros/humble/include/rclcpp/rclcpp/client.hpp:550:10: error: no member named 'optional' in namespace 'std' [clang-diagnostic-error]
>     std::optional<CallbackInfoVariant>
>     ~~~~~^
> /opt/ros/humble/include/rclcpp/rclcpp/client.hpp:550:19: error: use of undeclared identifier 'CallbackInfoVariant' [clang-diagnostic-error]
>     std::optional<CallbackInfoVariant>
>                   ^
> ...
> ```

Working on a commit to fix this. One cause of the error is because ROS humble requires C++17 or greater but we were building for C++14. I'll let you know when I resolve these errors.

Reply:
I pushed commits to this branch and the v2-updates branch of network_systems to fix these issues. PR for network_systems is here: https://github.com/UBCSailbot/network_systems/pull/10

Make sure you updates src/new_project.repos so that it targets the v2-updates branch of network_systems if you want it to work.

Reply:
It built but print something to stderr:
```
--- stderr: network_systems                                
Building Network Systems with build type 'RelWithDebInfo' and flags: '-O2 -g -DNDEBUG -std=c++20 -Wall -Wextra -Wpedantic -Werror -Wno-unused-parameter -pthread'
```

Is this an error? If not, is there a way not to print this?

Reply:
> It built but print something to stderr:
> 
> ```
> --- stderr: network_systems                                
> Building Network Systems with build type 'RelWithDebInfo' and flags: '-O2 -g -DNDEBUG -std=c++20 -Wall -Wextra -Wpedantic -Werror -Wno-unused-parameter -pthread'
> ```
> 
> Is this an error? If not, is there a way not to print this?

Not an error. I set it up that way because printing to stdout doesn't work for some reason.